File: 01-programming-languages/Kotlin/01-fundamentals/02-control-flow/control-flow.kt
// Kotlin Control Flow - if, when, for, while

fun main() {
    // if expression
    val x = 10
    val y = 20
    
    // if as expression
    val max = if (x > y) x else y
    println("Max: $max")
    
    // if with blocks
    val result = if (x > y) {
        println("x is greater")
        x
    } else {
        println("y is greater")
        y
    }
    println("Result: $result")
    
    // when expression (like switch)
    val day = "Monday"
    val dayType = when (day) {
        "Monday", "Tuesday", "Wednesday", "Thursday", "Friday" -> "Weekday"
        "Saturday", "Sunday" -> "Weekend"
        else -> "Unknown"
    }
    println("$day is a $dayType")
    
    // when with conditions
    val temperature = 25
    val weather = when {
        temperature > 30 -> "Hot"
        temperature > 20 -> "Warm"
        temperature > 10 -> "Cool"
        else -> "Cold"
    }
    println("Weather: $weather")
    
    // when with any type
    val obj: Any = 42
    val description = when (obj) {
        is Int -> "Integer: $obj"
        is String -> "String: $obj"
        is Boolean -> "Boolean: $obj"
        else -> "Unknown type"
    }
    println(description)
    
    // for loop
    print("For loop: ")
    for (i in 1..5) {
        print("$i ")
    }
    println()
    
    // for with step
    print("With step: ")
    for (i in 0..10 step 2) {
        print("$i ")
    }
    println()
    
    // for with downTo
    print("Down to: ")
    for (i in 10 downTo 1) {
        print("$i ")
    }
    println()
    
    // for with until
    print("Until: ")
    for (i in 0 until 5) {
        print("$i ")
    }
    println()
    
    // forEach
    print("ForEach: ")
    (1..5).forEach { print("$it ") }
    println()
    
    // while loop
    print("While: ")
    var count = 0
    while (count < 5) {
        print("$count ")
        count++
    }
    println()
    
    // do-while loop
    print("Do-while: ")
    count = 0
    do {
        print("$count ")
        count++
    } while (count < 5)
    println()
    
    // break and continue
    print("Break at 3: ")
    for (i in 1..5) {
        if (i == 3) break
        print("$i ")
    }
    println()
    
    print("Skip even: ")
    for (i in 1..10) {
        if (i % 2 == 0) continue
        print("$i ")
    }
    println()
    
    // Labeled breaks
    print("Labeled break: ")
    outer@ for (i in 1..3) {
        for (j in 1..3) {
            if (j == 2) continue@outer
            print("($i,$j) ")
        }
    }
    println()
    
    println("Control flow example running")
}

File: 01-programming-languages/Scala/01-fundamentals/02-control-flow/control-flow.scala
// Scala Control Flow

object ControlFlow {
  def main(args: Array[String]): Unit = {
    // if/else
    val x = 10
    val result = if (x > 5) "big" else "small"
    println(s"x is $result")

    // for loop
    println("for loop:")
    for (i <- 1 to 5) {
      print(s"$i ")
    }
    println()

    // for with yield
    val squares = for (i <- 1 to 5) yield i * i
    println(s"squares: $squares")

    // for with guard
    val evens = for (i <- 1 to 10 if i % 2 == 0) yield i
    println(s"evens: $evens")

    // while loop
    println("while loop:")
    var i = 0
    while (i < 5) {
      print(s"$i ")
      i += 1
    }
    println()

    // do-while
    println("do-while loop:")
    var j = 0
    do {
      print(s"$j ")
      j += 1
    } while (j < 5)
    println()

    // Pattern matching
    val day = "Monday"
    val dayType = day match {
      case "Monday" | "Tuesday" | "Wednesday" | "Thursday" | "Friday" => "Weekday"
      case "Saturday" | "Sunday" => "Weekend"
      case _ => "Unknown"
    }
    println(s"$day is a $dayType")

    // break and continue (via scala.util.control)
    import scala.util.control.Breaks._
    breakable {
      for (i <- 1 to 10) {
        if (i > 5) break
        print(s"$i ")
      }
    }
    println()
  }
}
